src/core/net/ip6.cpp

```diff
@@ -328,6 +333,24 @@ void Ip6::ProcessReceiveCallback(Message &aMessage)
 
     VerifyOrExit(sReceiveIp6DatagramCallback != NULL, ;);
 
+    if (sFilterControlTraffic)
+    {
+        Header ip6;
+        aMessage.Read(0, sizeof(ip6), &ip6);
+
+        // do not pass messages sent to/from an RLOC
+        VerifyOrExit(!ip6.GetSource().IsRoutingLocator() && !ip6.GetDestination().IsRoutingLocator(), ;);
+
+        if (ip6.GetNextHeader() == kProtoUdp)
```

Comment:
I think it makes sense to scope this further as you suggest.  MLE only uses the link-local address formed from its extended address.  For simplicity, we could add a condition that checks for link-local scope here.

A full and complete check would require would require checking against the link-local address used by one or more MLE instances tied to this IPv6 network component.